File: TRAB1/src/Votacao.java
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

public class Votacao {
    private LocalDate dataEleicao;
    private Map <Integer,Partido> partidos;
    private LinkedList<Candidato> candidatosTotais;
    private LinkedList<Candidato> candidatosEleitos;
    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
    
    public Votacao(){
        this.dataEleicao = null;
        this.partidos = new HashMap<>();
        this.candidatosTotais = new LinkedList<>();
        this.candidatosEleitos = new LinkedList<>();
    }

    public void setDataEleicao(String data) {
        this.dataEleicao = LocalDate.parse(data,formatter);
    }
    
    public void atualizaVotacaoPartido(Partido p, int chaveP){
        this.partidos.putIfAbsent(chaveP, p);
        
    }
    
    public void atualizaVotacaoCandidato(Candidato c){
        this.candidatosTotais.add(c);
        if(c.foiEleito()) this.candidatosEleitos.add(c);
        int idade = (int) ChronoUnit.YEARS.between(c.getNasc(), dataEleicao);
        c.setIdade(idade);       
    }
    
    public void apuraVotos(int numUrna, int quantVotos, String nomePart){
       
        if(numUrna>9 && numUrna<100){
            Partido p = this.partidos.get(numUrna);
            if(p!=null){
                p.incrementaVotosLegenda(quantVotos);
            }
            
        }
        if(numUrna>9999 && numUrna<100000){
            for(Candidato c: this.candidatosTotais){
                if(c.getNumUrna()==numUrna){
                    c.incrementaVotosCandidato(quantVotos);
                    break;
                }
            }
        }
    }
    
    public Partido getPartidoKey(int key){
        return this.partidos.get(key);
    }

    public void encontraEleitos(){
        for(Candidato c: candidatosTotais){
            if(c.foiEleito()) candidatosEleitos.add(c);
        }  
    }

    public int getQtdEleitos(){
        return this.candidatosEleitos.size();
    }

    public int getTotalVotosNominais(){
        int votosNominais = 0;
        for(Candidato c: this.candidatosTotais){
            votosNominais += c.getNumVotos();
        }
        return votosNominais;
    }
    public int foiMaisVotado(Candidato c){
        LinkedList <Candidato> todos = this.candidatosTotais;
        
        int i = 1;
        for(Candidato cand: todos){
            if(cand == c && i<=getQtdEleitos()) return -1;
            if(cand == c && i>getQtdEleitos()) break;
            i++;
        }
        return i;
    }
    
    public int getTotalVotosLegenda(){
        int votosLegenda = 0;
        for(Partido p: this.partidos.values()){
            votosLegenda += p.getQtdVotosLegenda();
        }
        return votosLegenda;
    }

    public boolean contemPartido(int key){
        return (this.partidos.containsKey(key));
    }

    public LinkedList<Candidato> getCandidatosTotais(){
        return new LinkedList<>(this.candidatosTotais);
    }

    public LinkedList<Candidato> getCandidatosEleitos(){
        return new LinkedList<>(this.candidatosEleitos);
    }
    
    public HashMap<Integer,Partido> getPartidos(){

        return new HashMap<>(this.partidos);
    }
    
    public boolean foiEleito(Candidato c){
        for(Candidato cEleito: this.candidatosEleitos){
            if(c==cEleito) return true;
        }
        return false;
    }
    public void ordenaCandidatos(){
        Collections.sort(this.candidatosTotais, new CandidatoComparator());
        Collections.sort(this.candidatosEleitos, new CandidatoComparator());
    }
}

File: TRAB2/src/Votacao.cpp
#include "Votacao.hpp"

Votacao::Votacao() : dataEleicao(), candidatosTotais(), candidatosEleitos(), partidos() {}

//MARCELA: se mudar o tipo da data, pode transfomar de string para o tipo aqui mesmo, ao invés de lidar com isso fora daqui
void Votacao::setDataEleicao(string &data){
    this->dataEleicao = data;
}

void Votacao::atualizaVotacaoPartido(Partido *p, int chaveP) {
    this->partidos.insert(make_pair(chaveP, p));
}

void Votacao::atualizaVotacaoCandidato(Candidato *c){
    this->candidatosTotais.push_back(c);
    if(c->verificaSeFoiEleito()) this->candidatosEleitos.push_back(c);
    c->calculaIdade(dataEleicao);
}      

void Votacao::apuraVotos(int numUrna, int quantVotos, string &nomePart){
    if(numUrna>9 && numUrna<100){
        auto it = this->partidos.find(numUrna);
        if (it != this->partidos.end()) {
            it->second->incrementaVotosLegenda(quantVotos);
        }
    }
    
    if(numUrna>9999 && numUrna<100000){
        for(Candidato *c: this->candidatosTotais){
            if(c->getNumUrna()==numUrna){
                c->incrementaVotosCandidato(quantVotos);
                break;
            }
        }
    }
}

const Partido *Votacao::getPartidoKey(int key) const{
    auto it = this->partidos.find(key);
    if (it == this->partidos.end()) return nullptr;
    else return it->second; 
}

int Votacao::getQtdEleitos() const{
    return this->candidatosEleitos.size();
}

int Votacao::getTotalVotosNominais()const{
    int qtdNominais = 0;
    for(Candidato *c: this->candidatosTotais){
        qtdNominais += c->getNumVotos();
    }
    return qtdNominais;
}

int Votacao::foiMaisVotado(Candidato *c) const{
    int i = 1;
    for(Candidato *cand: this->candidatosTotais){
        if(cand == c && i<=getQtdEleitos()) return -1;
        if(cand == c && i>getQtdEleitos()) break;
        i++;
    }
    return i;
} 

int Votacao::getTotalVotosLegenda() const{
    int votosLegenda = 0;
    for (const auto &par : this->partidos) {
        votosLegenda += par.second->getQtdVotosLegenda();
    }
    return votosLegenda;
}

bool Votacao::contemPartido(int key) const{
    return this->partidos.find(key) != this->partidos.end();
}

list<Candidato*> Votacao::getCandidatosTotais() const{
    list<Candidato*> copia = candidatosTotais;
    return copia;
}

list<Candidato*> Votacao::getCandidatosEleitos() const{
    list<Candidato*> copia = candidatosEleitos;
    return copia;
}   

map<int,Partido*> Votacao::getPartidos(){
    map<int, Partido*> copia = this->partidos;
    return copia;
}

bool Votacao::foiEleito(Candidato *c) const{
    for(Candidato *cEleito: this->candidatosEleitos){
        if(c==cEleito) return true;
    }
    return false;
}

void Votacao::ordenaCandidatos(){
    Comparator comp;

    this->candidatosTotais.sort([&comp](Candidato* a, Candidato* b) {
        return comp.comparaCandidatos(*a, *b);
    });
    
    this->candidatosEleitos.sort([&comp](Candidato* a, Candidato* b) {
        return comp.comparaCandidatos(*a, *b);
    });
}